cmd/clusterctl/client/config/imagemeta_client.go

```diff
@@ -0,0 +1,136 @@
+}
+
+// imageMetaClient implements ImageMetaClient.
+type imageMetaClient struct {
+	reader         Reader
+	imageMetaCache map[string]*imageMeta
+}
+
+// ensure imageMetaClient implements ImageMetaClient.
+var _ ImageMetaClient = &imageMetaClient{}
+
+func newImageMetaClient(reader Reader) *imageMetaClient {
+	return &imageMetaClient{
+		reader:         reader,
+		imageMetaCache: map[string]*imageMeta{},
+	}
+}
+
+func (p *imageMetaClient) AlterImage(component, image string) (string, error) {
+	// Gets the image meta that applies to the selected component; if none, returns early
+	meta, err := p.getImageMetaByComponent(component)
+	if err != nil {
+		return "", err
+	}
+	if meta == nil {
+		return image, nil
+	}
+
+	// Apply the image meta to image name
+	return meta.ApplyToImage(image)
+}
+
+// getImageMetaByComponent returns the image meta that applies to the selected component
+func (p *imageMetaClient) getImageMetaByComponent(component string) (*imageMeta, error) {
+	// if the image meta for the component is already known, return it
+	if im, ok := p.imageMetaCache[component]; ok {
+		return im, nil
+	}
+
+	// Otherwise read the image override configurations.
+	var meta map[string]imageMeta
+	if err := p.reader.UnmarshalKey(imagesConfigKey, &meta); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal image override configurations")
+	}
+
+	// If there are not image override configurations, return.
+	if meta == nil {
+		p.imageMetaCache[component] = nil
+		return nil, nil
+	}
+
+	// Gets the image configuration and to the specific component, and returns the union of the two.
+	m := &imageMeta{}
+	if allMeta, ok := meta[allImageConfig]; ok {
+		m.Union(allMeta)
+	}
+	if componentMeta, ok := meta[component]; ok {
+		m.Union(componentMeta)
+	}
+
+	p.imageMetaCache[component] = m
+	return m, nil
+}
+
+// imageMeta allows to define transformations to apply to the image contained in the YAML manifests.
+type imageMeta struct {
+	// repository sets the container registry to pull images from.
+	Repository string `json:"repository,omitempty"`
+
+	// Tag allows to specify a tag for the images.
+	Tag string `json:"tag,omitempty"`
+}
+
+// Union allows to merge two imageMeta transformation; in case both the imageMeta defines new values for the same field,
+// the other transformation takes precedence on the existing one.
+func (i *imageMeta) Union(other imageMeta) {
+	if other.Repository != "" {
+		i.Repository = other.Repository
+	}
+	if other.Tag != "" {
+		i.Tag = other.Tag
+	}
+}
+
+// ApplyToImage changes an image name applying the transformations defined in the current imageMeta.
+func (i *imageMeta) ApplyToImage(image string) (string, error) {
+	// Splits the image name into its own composing parts
+	imageSplit, err := util.NewImageSplit(image)
+	if err != nil {
+		return "", err
+	}
+
+	// apply transformations
+	if i.Repository != "" {
+		imageSplit.Repository = i.Repository
+	}
+	if i.Tag != "" {
+		imageSplit.Tag = i.Tag
+	}
```

Comment:
the above looks similar to Union(), potentially there could be a private union() that Union() and this code call.

Reply:
they look similar, but in this case the underlying types are different because we are comparing and imageMeta and an imageSplit